### ISSUE TITLE
[BUG]: The Chronicles of Narnia - The Lion, The Witch and The Wardrobe: Heat haze rendering issue in software mode

### ISSUE THREAD

**Issue by Silanda** [PCSX2/pcsx2#4836]
### Describe the Bug

This is hard to describe but quite easy to see in motion. 

At a few places in the game there are fires which cause a heat haze effect (wavy distortion of objects behind the fire). Using the software renderer, if you move while the effect is on screen, instead of just distorting objects behind where the effect is being rendered it can look like afterimages are being "dragged" into the effect. Sometimes parts of the backgrounds to the side of where the effect is being rendered are incorrectly displayed in the heat haze where they shouldn't be, and sometimes it appears that parts of the background covered by the effect remain static for a frame or so during constant movement, leading to a double image effect.

This does not happen with the hardware renderers and did not appear to happen on actual PS2 hardware when I tested it.

### Reproduction Steps

Play the game to the point where there's a fire. Move around while observing the heat haze effect. The second level is the first instance where the issue manifests.

### Expected Behavior

The heat haze should look like heat haze without glitchy afterimages. 

### PCSX2 Revision

v1.7.0-dev-1814-g71b2c17e2

### Operating System

Windows 10 (64bit)

### CPU

Ryzen 3700X

### GPU

RTX 3070

### GS Settings

None have any impact as far as I can tell.

### Emulation Settings

None have any impact as far as I can tell.

### GS Window Screenshots

Hard to capture in single screenshots but it is just about visible here:

![Narnia1](https://user-images.githubusercontent.com/4227434/134815766-f0658459-c354-4ab7-92ab-f47a2cfb9a48.png)
![Narnia2](https://user-images.githubusercontent.com/4227434/134815769-3d16698d-129b-42e1-bc73-767ae72b2e8f.png)

### Logs & Dumps

Two GSDumps. The first is a multi-frame dump where the issue is visible if stepped through. The second is a single frame dump where the issue as described isn't visible, but something odd is definitely going on when rendered in software mode.

https://mega.nz/file/4fBhyC7b#dROUBEcUWa0qgYoNtKuJwuhlH_IFrSemZP0qG-i0UOk
[Narnia2.zip](https://github.com/PCSX2/pcsx2/files/7231719/Narnia2.zip)

**Comment by refractionpcsx2** [comment]
Does this happen in HW mode too?

**Comment by Silanda** [comment]
No. Only software mode.

**Comment by prafullpcsx2** [comment]
@Silanda Can you download r5611 and r5612 from here to confirm it's r5612 that breaks SW mode rendering in this game?

https://buildbot.orphis.net/pcsx2svn/index.php?m=fulllist

**Comment by Silanda** [comment]
I can confirm that was the change that broke it. Nicely found!

**Comment by refractionpcsx2** [comment]
I did some investigation this evening, baring in mind my knowledge of the GS code is *very* limited.

But there is an optimisation which was enabled in the previously mentioned commit, which looks to be basically incompatible with the software renderer.

Both this game and Test Drive (which also has SW issues due to that commit) both try to use CLAMP mode clamping and that optimisation basically tries to cut out just the part of the texture which is going to get read during the current draw, the problem is, I think the software renderer is expecting the whole texture, so it completely throws it off.

I could be wrong on this, but stopping that optimisation running when in SW mode and using CLAMP mode clamping does correct the fire problem (and test drive).

I've let @tadanokojin know what I've discovered too, so he has a clue what he's looking for when he gets around to looking at it (since he's better at this than me :D )
